test/integration/image-future/default/test/index.test.ts

```diff
@@ -387,6 +387,14 @@ function runTests(mode) {
     )
   })
 
+  it('should callback native onError even when error before hydration', async () => {
+    let browser = await webdriver(appPort, '/on-error-before-hydration')
+    await check(
+      () => browser.eval(`document.getElementById("msg").textContent`),
+      'error state'
+    )
+  })
+
```

Comment:
This test doesn't really assert anything about which branch of the race condition is being tested. Was a similar test failing 100% of the time before?

Can you make it so next's hydration waits until after the img has errored explicitly?

Reply:
The test title says “before hydration” and the page name has before-hydration. 

I also confirmed this reproduces every time locally. 

I don’t think we can write a e2e test that delays Next hydration during “next start”

Reply:
We could potentially add a test method similar to the `__NEXT_HYDRATED_CB` we use to delay hydrating until a Promise is resolved

Reply:
>The test title says “before hydration” and the page name has before-hydration.

Right I meant that the test suggests there is an error before hydration but it does not guarantee the error happened before hydration so in the future if we deleted this fix accidentally the test could very well continue to pass even though it should start failing.

>I also confirmed this reproduces every time locally.

So if this is the case I suppose it's ok but that could potentially change in the future depending on where the tests are run or how Next or React changes. It get's a little heavy handed I suppose to try to contrive these things but the test would feel a lot more valuable if you knew it was fail 100% of the time if the fix was removed.

Reply:
Yeah if react could fetch, parse, and hydrate before the `<img>` is fetched and erros, its possible it might not exercise the new code path. I'll see if we can delay hydration, its just something we can do with a production build